tree/treeplayer/src/TTreeProcessorMT.cxx

```diff
@@ -37,6 +37,92 @@ struct EntryCluster {
    Long64_t end;
 };
 
+// note that this routine assumes global entry numbers
+static bool ClustersAreSortedAndContiguous(const std::vector<std::vector<EntryCluster>> &cls)
+{
+   Long64_t last_end = 0ll;
+   for (const auto &fcl : cls) {
+      for (const auto &c : fcl) {
+         if (last_end != c.start)
+            return false;
+         last_end = c.end;
+      }
+   }
+   return true;
+}
+
+/// Take a vector of vectors of EntryClusters (a vector per file), filter the entries according to entryList, and
+/// and return a new vector of vectors of EntryClusters where cluster start/end entry numbers have been converted to
+/// TEntryList-local entry numbers.
+///
+/// This routine assumes that entry numbers in the TEntryList (and, if present, in the sub-entrylists) are in
+/// ascending order, i.e., for n > m:
+///   elist.GetEntry(n) + tree_offset_for_entry_from_elist(n) > elist.GetEntry(m) + tree_offset_for_entry_from_elist(m)
+static std::vector<std::vector<EntryCluster>>
+ConvertToElistClusters(std::vector<std::vector<EntryCluster>> &&clusters, TEntryList &entryList,
+                       const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames,
+                       const std::vector<Long64_t> &entriesPerFile)
+{
+   R__ASSERT(entryList.GetN() > 0); // wasteful to call this function if it has nothing to do
+   R__ASSERT(ClustersAreSortedAndContiguous(clusters));
+
+   const bool listHasGlobalEntryNumbers = entryList.GetLists() == nullptr;
+   const auto nFiles = clusters.size();
+
+   std::unique_ptr<TChain> chain;
```

Comment:
```suggestion
   TChain chain;
```

Reply:
I think a `unique_ptr` makes it clear that we might not need the `TChain` depending on the situation

Reply:
fair enough.